### PR TITLE
Expose the WeatherOverlay to lua

### LINKED ISSUE #11538: Lua Scripted WeatherOverlay Commands
It would be nice if you can control the WeatherOverlay trait through Lua. You can already control the Lighting with Lua.

### PR DESCRIPTION
**Pull request by abcdefg30**
Closes #11538.

A testcase can be found at my [lwTest branch](https://github.com/abcdefg30/OpenRA/tree/lwTest).

### CODE REVIEW

File: OpenRA.Mods.Common/Scripting/Global/WeatherGlobal.cs [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+				if (!hasAnOverlay)
+					return;
+
+				overlay.ParticleDensityFactor = (int)value;
+				overlay.CalculateMaxParticleCount();
+			}
+		}
+
+		[Desc("Should the level of the wind change over time, or just stick to the first value of WindLevels?")]
+		public bool ChangingWindLevel
```
Comment (pchote) [inline OpenRA.Mods.Common/Scripting/Global/WeatherGlobal.cs]: Do we really need to expose all this to lua?  Scripts can modify things themselves if they want it to evolve.

**Comment by pchote** [comment]
Could we use upgrades to switch between different instances of the weather trait defined in map.yaml?

File: OpenRA.Mods.Common/Scripting/Global/WeatherGlobal.cs (hunk 1 above)
Comment (pchote) [inline OpenRA.Mods.Common/Scripting/Global/WeatherGlobal.cs]: Directly mapping the trait's properties into lua will make it much more difficult if we want to change the way the trait works in the future (which we know that we need to, e.g. wind).

**Comment by abcdefg30** [comment]
> Could we use upgrades to switch between different instances of the weather trait defined in map.yaml?

This won't give us the accurancy one might want, or would force people into having hundreds of `WeatherOverlay` definition for a soft change from e.g. heavy rain to almost no rain, wouldn't it?

**Comment by pchote** [comment]
The weather trait can implement upgrade enabled/disabled handlers to fade the effect in/out to hide the transition between states.

**Comment by pchote** [comment]
Repeating/paraphrasing here since the logs aren't available for context:

My main objection is that this goes against our long-standing rule that Lua APIs should be task oriented and not directly bridge trait properties.  This is especially true for the weather trait because (a) its properties are low level and a bit horrible, and (b) we already know that these need to change (extracting wind out so that it can also affect gas projectiles in TS), and baking the current trait setup into our nominally-stable Lua API makes that much harder.

**Comment by abcdefg30** [comment]
Alright, so the next thing to get this through would be extracting wind to a different trait, and then query this new trait inside the `WeatherOverlay`?

I would extract following fields:
- `bool ChangingWindLevel`
- `int[] WindLevels`
- `int[] WindTick`
- `bool InstantWindChanges`

**Comment by Mailaender** [comment]
I assume https://github.com/OpenRA/OpenRA/issues/11641 was the dependency.

**Comment by Mailaender** [comment]
I have to agree with @pchote that keeping properties at trait level is easier to use and can be automatically updated when the weather code gets refactored. It would be better to have the WeatherOverlay upgradeable instead.

**Comment by abcdefg30** [comment]
> I assume #11641 was the dependency.

Actually this is the depedency:
> Alright, so the next thing to get this through would be extracting wind to a different trait, and then query this new trait inside the `WeatherOverlay`?

**Comment by Mailaender** [comment]
Even after a potential refactor for Tiberian Sun we don't really want this directly exposed to Lua and turn on/off traits functionality already exists with upgrades so just make the weather trait upgradeable. For extra polish you can implement the transitioning during upgrades.

**Comment by Mailaender** [comment]
That would also a first step towards https://github.com/OpenRA/ra2/issues/149.

**Comment by GraionDilach** [comment]
Uhm, I don't see how the Weather Control Device or ion storms (on-map effects) are related to our current weather overlay (whole-screen-covering visual). In fact, RA2 WCD shouldn't even use WeatherOverlay at all.

**Comment by Mailaender** [comment]
I thought about adding not just a palette effect upgrade, but also a stormy weather upgrade triggered by the Weather Control Device which would give it more polish.

**Comment by abcdefg30** [comment]
Closing this for now, as I won't have time for it in the next week.

**Pull request closed without merging**